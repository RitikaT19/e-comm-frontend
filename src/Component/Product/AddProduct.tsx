import React, { useState } from "react";
import cross from "../../assets/Icons/cross.png";
import { TextArea } from "../common/TextArea/TextArea";
import { Textfield } from "../common/Textfield/Textfield";
import "../styles/add-product.css";
// import Select from "react-select";
import Button from "../common/Button/Button";
import { Loading } from "../common/Loading/Loading";
// import "react-dropdown/style.css";
import {
  isQuantityValid,
  isPriceValid,
  isProductValid,
} from "../../helpers/validate";
import { DropDown } from "../common/DropDown/DropDown";

interface Props {
  CrossIconClick: any;
  submitProduct: any;
  showLoader: boolean;
  categoryNames: any;
  isEdit: boolean;
  productInformation: any;
  handleEditProduct: any;
  errorMessage: any;
  successMessage: any;
}

export const AddProduct: React.FC<Props> = ({
  CrossIconClick,
  handleEditProduct,
  submitProduct,
  showLoader,
  categoryNames,
  isEdit,
  productInformation,
  errorMessage,
  successMessage,
}) => {
  // stores name
  const [name, setName] = useState<string>(
    isEdit ? productInformation.name : ""
  );
  // stores name error
  const [nameError, setNameError] = useState<string>("");
  // stores price
  const [price, setPrice] = useState<string>(
    isEdit ? productInformation.price : ""
  );
  // stores price error
  const [priceError, setPriceError] = useState<string>("");
  // stores quantity
  const [quantity, setQuantity] = useState<string>(
    isEdit ? productInformation.quantity : ""
  );
  const [quantityError, setQuantityError] = useState<string>("");
  // stores description
  const [description, setDescription] = useState<string>(
    isEdit ? productInformation.description : ""
  );

  const [selectedCategory, setSelectedCategory] = useState<string>(
    isEdit ? productInformation.category : ""
  );

  // stores empty field error
  const [showEmptyFieldError, setShowEmptyFieldError] = useState(false);

  // function to verify product name
  const checkName = async () => {
    const verifyNameResult = isProductValid(name);
    if (!verifyNameResult) {
      setNameError("Enter a valid product");
    } else {
      setNameError("");
    }
  };

  // verify quantity
  const checkQuantity = async () => {
    const verifyQuantityResult = isQuantityValid(quantity);
    if (!verifyQuantityResult) {
      setQuantityError("Enter valid quantity");
    } else {
      setQuantityError("");
    }
  };

  // verify price
  const checkPrice = async () => {
    const verifyPriceResult = isPriceValid(price);
    if (!verifyPriceResult) {
      setPriceError("Enter valid Price");
    } else {
      setPriceError("");
    }
  };

  // function for when onClickProduct is clicked
  const onClickProduct = async () => {
    if (!(name || quantity || selectedCategory || price || description)) {
      setShowEmptyFieldError(true);
    } else {
      await submitProduct({
        name,
        price,
        quantity,
        description,
        category: selectedCategory,
      });
    }
  };

  const onEditProduct = async () => {
    await handleEditProduct({
      id: productInformation?._id,
      name,
      price,
      quantity,
      description,
      category: selectedCategory,
    });
  };

  let options: any = [];
  categoryNames.map((category: any, index: number) => {
    options.push({ value: category._id, label: category.name });
  });

  console.log(options, "category names");
  return (
    <div>
      {showLoader ? (
        <Loading />
      ) : (
        <div className="add-product-main-container">
          <div className="add-product-div">
            <div className="close-button">
              <img src={cross} alt="cross" onClick={CrossIconClick} />
            </div>
            <label htmlFor="text">Category</label>
            <div className="dropdown-category">
              <DropDown
                label="Select category"
                options={options}
                id="category-dropdown"
                handleChange={(e: any) => setSelectedCategory(e.value)}
                showLabel={false}
                value={selectedCategory}
              />
            </div>
            {/* textfield for name */}
            <Textfield
              label="Name"
              placeholder="Enter the name of the product"
              value={name}
              id="name"
              onChange={(e: any) => setName(e.target.value)}
              error={nameError}
              onBlur={checkName}
            />
            {/*textfield for price  */}
            <Textfield
              label="Price"
              placeholder="Enter the price of the product"
              value={price}
              id="price"
              onChange={(e: any) => setPrice(e.target.value)}
              error={priceError}
              onBlur={checkPrice}
            />

            {/* Textfield for quantity */}
            <Textfield
              label="Quantity"
              placeholder="Enter the quantity of the product"
              value={quantity}
              id="quantity"
              onChange={(e: any) => setQuantity(e.target.value)}
              error={quantityError}
              onBlur={checkQuantity}
            />
            {/* textarea for description */}
            <TextArea
              label="Description"
              placeholder="Enter the description of the product"
              value={description}
              id="description"
              onChange={(e: any) => setDescription(e.target.value)}
            />
            {errorMessage ? (
              <p className="sign-in-error">{errorMessage}</p>
            ) : successMessage ? (
              <p className="sign-in-success">{successMessage}</p>
            ) : (
              showEmptyFieldError && (
                <p className="sign-in-error">Please fill all the fields</p>
              )
            )}

            {/* Submit button for adding product */}
            <Button
              id={isEdit ? "edit-product-button" : "add-new-product-button"}
              value={isEdit ? "Edit Product" : "Add Product"}
              handleClick={isEdit ? onEditProduct : onClickProduct}
            />
          </div>
        </div>
      )}
    </div>
  );
};
