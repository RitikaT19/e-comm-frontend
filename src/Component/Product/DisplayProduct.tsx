import React from "react";
import productImg from "../../assets/Icons/mobile.jpeg";
import Button from "../common/Button/Button";
import "../styles/display-product.css";
import { Loading } from "../common/Loading/Loading";

// define required props by the component
interface Props {
  productDetails: any;
  deleteProduct: any;
  showLoader: boolean;
  updateProductIconClick: any
}
export const DisplayProduct: React.FC<Props> = ({
  productDetails,
  deleteProduct,
  showLoader,
  updateProductIconClick
}) => {
  const editProduct = async(item: any) =>{
    
  };
  const onDeleteProduct = async (index: number) => {
    deleteProduct(productDetails[index]?._id, productDetails[index]?.category);
  };
  return (
    <div>
      {showLoader ? (
        <Loading />
      ) : (
        <div className="display-product-main-container">
          <div>
            {/* check the length of the productDetails array */}
            {productDetails.length > 0 ? (
              <div className="list-product-details">
                {/* map product details */}
                {productDetails.map((item: any, index: number) => (
                  <div className="product-card">
                    <div key={index}>
                      {/* product image */}
                      <img
                        className="image"
                        src={productImg}
                        alt="product image"
                      />
                      {/* product name */}
                      <p id={"product-name"}>{item.name}</p>
                      {/* product price */}
                      <p id={"product-price"}>
                        <b>Price: </b>Rs.{item.price}
                      </p>
                      {/* product quantity */}
                      <p id={"display-category-quantity"}>
                        <b>Quantity:</b>
                        {item.quantity}
                      </p>
                      {/* product description */}
                      <p id={"display-category-description"}>
                        <b>Description: </b>
                        {item.description}
                      </p>
                      <div className="buttons-div">
                        {/* Button for editing product */}
                        <Button
                          id="edit-product-button"
                          value="Edit"
                          handleClick={()=>updateProductIconClick(item)}
                        />
                        {/* product for deleting category */}
                        <Button
                          id="delete-product-button"
                          value="Delete"
                          handleClick={() => onDeleteProduct(index)}
                        />
                      </div>
                    </div>
                  </div>
                ))}
              </div>
            ) : null}
          </div>
        </div>
      )}
    </div>
  );
};
