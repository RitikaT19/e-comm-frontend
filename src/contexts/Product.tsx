import React, { useReducer, createContext } from "react";
import {
  Product,
  initialState,
  State as InitialStateType,
  Actions,
} from "../reducers/product";

export const ProductContext = createContext<{
  state: InitialStateType;
  dispatch: React.Dispatch<Actions>;
}>({
  state: initialState,
  dispatch: () => null,
});

export const ProductContextProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(Product, initialState);
  const value = { state, dispatch };
  return (
    <ProductContext.Provider value={value}>{children}</ProductContext.Provider>
  );
};
