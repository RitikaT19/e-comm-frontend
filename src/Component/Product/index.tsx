import React, { useContext, useEffect, useState } from "react";
import { SideBar } from "../common/SideBar/Sidebar";
import {
  getProductById,
  createProduct,
  removeProduct,
  editProduct,
  clearErrors
} from "../../actions/product";
import { getCategory } from "../../actions/category";
import { CategoryContext } from "../../contexts/Category";
import { ProductContext } from "../../contexts/Product";
import { LoadingContext } from "../../contexts/Loading";
import { DisplayProduct } from "./DisplayProduct";
import { AddProduct } from "./AddProduct";
import "../styles/product.css";
import Button from "../common/Button/Button";
import { MenuHeader } from "../MenuHeader";
import { useParams } from "react-router";

export const Product: React.FC = () => {
  const { id }: any = useParams();
  //rename state and dispatch as productState and productDispatch respectively
  const { state: productState, dispatch: productDispatch } =
    useContext(ProductContext);
  //rename state and dispatch as categoryState and categoryDispatch respectively
  const { state: categoryState, dispatch: categoryDispatch } =
    useContext(CategoryContext);
  const { dispatch: loadingDispatch } = useContext(LoadingContext);
  // flag for showing add product page
  const [showProductPage, setShowProductPage] = useState(false);
  // flag for add product header
  const [showAddProductLoader, setShowAddProductLoader] = useState(false);
  // flag for display product header
  const [showDisplayProductLoader, setShowDisplayProductLoader] =
    useState(false);
  // stores the flag for the edit category
  const [editProductData, setEditProductData] = useState<boolean>(false);
  // stores the detail of a single product
  const [productData, setProductData] = useState<any>();

  // function for fetching products
  const fetchProducts = async (id: string) => {
    setShowDisplayProductLoader(true);
    // call getProductById action
    await getProductById(id)(productDispatch, loadingDispatch).then(() => {
      setShowDisplayProductLoader(false);
    });
  };

  // call useeffectt when getProductById is successful
  useEffect(() => {
    fetchProducts(id);
  }, [id]);

  // function for toggling add product page
  const addProductToggle = () => {
    setEditProductData(false);
    setProductData("");
    setShowProductPage(!showProductPage);
  };

  // function for adding a new product
  const addProduct = async (data: any) => {
    setShowAddProductLoader(true);
    // call createProduct action
    await createProduct(data)(productDispatch, loadingDispatch).then(() => {
      setShowAddProductLoader(false);
    });
    setTimeout(() => {
      setShowProductPage(false);
      clearErrors(productDispatch)
    }, 2000);
    // call fetchProducts after adding a new product
    fetchProducts(id);
  };

  // function for deleting product
  const deleteProduct = async (id: any, category: any) => {
    // call removeProduct action
    await removeProduct(id)(productDispatch, loadingDispatch);
    // call the fetchProducts after deleting a product
    fetchProducts(category);
  };

  // function for fetching category
  const fetchCategory = async () => {
    // call getCategory action
    await getCategory()(categoryDispatch, loadingDispatch).then(() => {});
  };

  // function for updating product
  const updateProduct = async (data: any) => {
    setShowAddProductLoader(true);
    await editProduct(data)(productDispatch, loadingDispatch).then(() => {
      setShowAddProductLoader(false);
    });
    setShowProductPage(false);
    fetchProducts(id);
  };

  const updateProductIconClick = (data: any) => {
    // sets the product detail for the card clicked
    setProductData(data);
    // set setShowProductPage as true
    setShowProductPage(true);
    // sets the edit flag as true
    setEditProductData(true);
  };
  return (
    <>
      <SideBar />
      <MenuHeader />
      <div className="actionButton">
        <p className="add-product-heading">Products</p>
        <Button
          id="add-product-button"
          value="Add Product"
          handleClick={addProductToggle}
        />
      </div>
      {showProductPage ? (
        <AddProduct
          CrossIconClick={addProductToggle}
          submitProduct={addProduct}
          showLoader={showAddProductLoader}
          isEdit={editProductData}
          productInformation={productData}
          categoryNames={categoryState.fetchCategorySuccess}
          handleEditProduct={updateProduct}
          errorMessage={
            editProductData
              ? productState.updateProductError
              : productState.error
          }
          successMessage={
            editProductData
              ? productState.updateProductSuccess
              : productState.updateProductError
          }
        />
      ) : (
        <DisplayProduct
          productDetails={productState.fetchProductSuccess}
          deleteProduct={deleteProduct}
          showLoader={showDisplayProductLoader}
          updateProductIconClick={updateProductIconClick}
        />
      )}
    </>
  );
};
