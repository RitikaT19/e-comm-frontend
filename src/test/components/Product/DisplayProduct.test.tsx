import React, { Component } from "react";
import { mount, ReactWrapper } from "enzyme";
import { DisplayProduct } from "../../../component/Product/DisplayProduct";

describe("Testing Display Products", () => {
  let wrapper: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>>;

  const deleteProduct: any = jest.fn();
  const showLoader: boolean = false;
  const updateProductIconClick: any = jest.fn();
  const productDetails = {
    name: "Test",
    quantity: "10",
    price: "10",
    category: "test",
    description: "test",
  };

  const defaultProps = {
    productDetails,
    deleteProduct,
    showLoader: false,
    updateProductIconClick,
  };

  beforeEach(() =>{
    wrapper = mount(<DisplayProduct{...defaultProps}/>)
})
it("should match snapshot",() =>{
    expect(wrapper).toMatchSnapshot()
})
it("should have 0 button component", () =>{
   expect(wrapper.find("Button")).toHaveLength(0)
})
it("should click delete and edit icon for a card", () => {
   expect(wrapper.find("#delete-product-button")).toBeTruthy();
   expect(wrapper.find("#edit-product-button")).toBeTruthy();
 });

});
