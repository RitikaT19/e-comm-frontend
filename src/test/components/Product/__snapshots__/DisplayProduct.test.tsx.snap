// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Display Products should match snapshot 1`] = `
<DisplayProduct
  deleteProduct={[MockFunction]}
  productDetails={
    Object {
      "category": "test",
      "description": "test",
      "name": "Test",
      "price": "10",
      "quantity": "10",
    }
  }
  showLoader={false}
  updateProductIconClick={[MockFunction]}
>
  <div>
    <div
      className="display-product-main-container"
    >
      <div />
    </div>
  </div>
</DisplayProduct>
`;
